File: ts/src/engine.ts
/**
 * The compiled engine: every CEL expression, regex pattern and Mustache
 * template of a schema compiled exactly once, up front. A bad schema fails
 * here as a SchemaError, never later at evaluation time. Mirrors the Go
 * reference engine.go.
 */

import {
  type CelEnv,
  type CelInput,
  type CelResult,
  CelScalar,
  type CelValue,
  celEnv,
  isCelError,
  isCelList,
  isCelMap,
  isCelUint,
  mapType,
  parse,
  plan,
} from "@bufbuild/cel";
import { strings } from "@bufbuild/cel/ext";
import type { Expr, ParsedExpr } from "@bufbuild/cel-spec/cel/expr/syntax_pb.js";
import { isMessage } from "@bufbuild/protobuf";
import { isReflectMessage } from "@bufbuild/protobuf/reflect";
import { DurationSchema, TimestampSchema } from "@bufbuild/protobuf/wkt";
import Mustache from "mustache";
import { checkDescriptor, joinPath, nestedSchemas, SchemaError, schemaErr } from "./descriptor.js";
import type { CompileOptions, FormatFunc, FormatRegistry } from "./format.js";
import { coreFormats } from "./format.js";
import type { ValidationError } from "./gen/schemapb/errors_pb.js";
import type { Schema, Schema_Field } from "./gen/schemapb/schema_pb.js";
import type { Format } from "./typed.js";
import type { Native, NativeStruct } from "./value.js";

type Vars = {
  this: typeof CelScalar.DYN;
  root: ReturnType<typeof mapType<typeof CelScalar.STRING, typeof CelScalar.DYN>>;
  index: typeof CelScalar.INT;
};

type Plan = (bindings: { this: CelInput; root: CelInput; index: bigint }) => CelResult;

/** Variable bindings for one evaluation. */
export interface EvalVars {
  this?: Native;
  root: NativeStruct;
  index?: bigint;
}

/** The result of one evaluation: a native value or an error message. */
export type EvalOutcome = { ok: true; value: Native } | { ok: false; error: string };

let sharedEnv: CelEnv<Vars> | undefined;

function env(): CelEnv<Vars> {
  sharedEnv ??= celEnv({
    variables: {
      this: CelScalar.DYN,
      root: mapType(CelScalar.STRING, CelScalar.DYN),
      index: CelScalar.INT,
    },
    funcs: strings,
  });
  return sharedEnv;
}

/** An immutable compiled schema. */
export class Engine {
  readonly schema: Schema;
  readonly formats: FormatRegistry;
  readonly regexps: Map<string, RegExp>;
  readonly #plans: Map<string, Plan>;
  readonly #asts: Map<string, ParsedExpr>;

  private constructor(
    schema: Schema,
    formats: FormatRegistry,
    regexps: Map<string, RegExp>,
    plans: Map<string, Plan>,
    asts: Map<string, ParsedExpr>,
  ) {
    this.schema = schema;
    this.formats = formats;
    this.regexps = regexps;
    this.#plans = plans;
    this.#asts = asts;
  }

  /**
   * Compiles a schema: descriptor checks, all expressions and patterns and
   * templates, static rejection of top-level computed cycles. Throws
   * SchemaError on any defect.
   */
  static compile(schema: Schema, opts: CompileOptions = {}): Engine {
    const errs: ValidationError[] = [...checkDescriptor(schema)];

    const formats = coreFormats();
    const extra =
      opts.formats instanceof Map
        ? opts.formats
        : new Map(Object.entries(opts.formats ?? {}) as [Format, FormatFunc][]);
    for (const [name, fn] of extra) {
      formats.set(name, fn);
    }

    const plans = new Map<string, Plan>();
    const asts = new Map<string, ParsedExpr>();
    for (const [src, path] of schemaExprs(schema)) {
      if (plans.has(src)) {
        continue;
      }
      try {
        const ast = parse(src);
        asts.set(src, ast);
        plans.set(src, plan(env(), ast) as Plan);
      } catch (e) {
        errs.push(schemaErr(path, `cel: ${e instanceof Error ? e.message : String(e)}`));
      }
    }

    const regexps = new Map<string, RegExp>();
    for (const [pattern, path] of schemaPatterns(schema)) {
      if (regexps.has(pattern)) {
        continue;
      }
      try {
        regexps.set(pattern, new RegExp(pattern, "u"));
      } catch (e) {
        errs.push(schemaErr(path, `pattern: ${e instanceof Error ? e.message : String(e)}`));
      }
    }

    for (const [name, src] of Object.entries(schema.templates)) {
      try {
        Mustache.parse(src);
      } catch (e) {
        errs.push(
          schemaErr(`templates.${name}`, `mustache: ${e instanceof Error ? e.message : String(e)}`),
        );
      }
    }

    const engine = new Engine(schema, formats, regexps, plans, asts);
    if (errs.length === 0) {
      errs.push(...engine.checkComputedCycles());
    }
    if (errs.length > 0) {
      throw new SchemaError(errs);
    }
    return engine;
  }

  /**
   * Runs a precompiled CEL plan (a sandboxed expression language — no access
   * to the JS runtime; this is NOT JavaScript eval). The result is in the
   * native value model.
   */
  eval(src: string, vars: EvalVars): EvalOutcome {
    const prg = this.#plans.get(src);
    if (prg === undefined) {
      return { ok: false, error: `expression not compiled: ${src}` };
    }
    const out = prg({
      this: nativeToCel(vars.this ?? null),
      root: nativeToCel(vars.root),
      index: vars.index ?? 0n,
    });
    if (isCelError(out)) {
      return { ok: false, error: out.message };
    }
    return { ok: true, value: celToNative(out) };
  }

  /** Runs a compiled expression and requires a boolean result. */
  evalBool(src: string, vars: EvalVars): { ok: boolean; error?: string } {
    const res = this.eval(src, vars);
    if (!res.ok) {
      return { ok: false, error: res.error };
    }
    if (typeof res.value !== "boolean") {
      return { ok: false, error: `expression yields ${typeof res.value}, want bool` };
    }
    return { ok: res.value };
  }

  /**
   * The dotted root paths a compiled expression reads (root.a -> "a",
   * root["a"] -> "a").
   */
  exprDeps(src: string): string[] {
    const ast = this.#asts.get(src);
    const root = ast?.expr;
    if (root === undefined) {
      return [];
    }
    const deps: string[] = [];
    walkExpr(root, deps);
    return deps;
  }

  /** Statically rejects cycles between top-level Computed fields. */
  private checkComputedCycles(): ValidationError[] {
    const computed = new Map<string, Schema_Field>();
    for (const f of this.schema.fields) {
      if (f.kind.case === "computed") {
        computed.set(f.name, f);
      }
    }
    if (computed.size === 0) {
      return [];
    }
    const deps = new Map<string, string[]>();
    for (const [name, f] of computed) {
      if (f.kind.case !== "computed") {
        continue;
      }
      deps.set(
        name,
        this.exprDeps(f.kind.value.expr).filter((d) => d !== name && computed.has(d)),
      );
    }
    const color = new Map<string, number>();
    const errs: ValidationError[] = [];
    const visit = (n: string): boolean => {
      const c = color.get(n) ?? 0;
      if (c === 1) {
        return false;
      }
      if (c === 2) {
        return true;
      }
      color.set(n, 1);
      for (const d of deps.get(n) ?? []) {
        if (!visit(d)) {
          return false;
        }
      }
      color.set(n, 2);
      return true;
    };
    for (const name of computed.keys()) {
      if (color.get(name) !== 2 && !visit(name)) {
        errs.push(schemaErr(name, "computed field cycle"));
      }
    }
    return errs;
  }
}

// =============================================================================
// Expression / pattern collection
// =============================================================================

function schemaExprs(s: Schema): Map<string, string> {
  const out = new Map<string, string>();
  const add = (src: string | undefined, path: string): void => {
    if (src !== undefined && src !== "" && !out.has(src)) {
      out.set(src, path);
    }
  };
  const walkSchema = (sub: Schema, prefix: string): void => {
    sub.rules.forEach((r, i) => {
      add(r.expr, `${prefix}#rule[${i}]`);
    });
  };
  const walkFields = (fields: Schema_Field[], prefix: string): void => {
    for (const f of fields) {
      const path = joinPath(prefix, f.name);
      add(f.when, `${path}#when`);
      add(f.normalize, `${path}#normalize`);
      f.rules.forEach((r, i) => {
        add(r.expr, `${path}#rule[${i}]`);
      });
      const kind = f.kind;
      if (kind.case === "computed") {
        add(kind.value.expr, `${path}#computed`);
      }
      if (kind.case === "choice") {
        add(kind.value.optionsExpr, `${path}#options`);
      }
      if (kind.case === "list") {
        add(kind.value.countExpr, `${path}#count`);
        walkFields(kind.value.items, `${path}[]`);
      }
      for (const child of nestedSchemas(f)) {
        walkSchema(child, path);
        walkFields(child.fields, path);
      }
    }
  };
  walkSchema(s, "");
  walkFields(s.fields, "");
  for (const [name, def] of Object.entries(s.defs)) {
    walkSchema(def, `$defs.${name}`);
    walkFields(def.fields, `$defs.${name}`);
  }
  return out;
}

function schemaPatterns(s: Schema): Map<string, string> {
  const out = new Map<string, string>();
  const walkFields = (fields: Schema_Field[], prefix: string): void => {
    for (const f of fields) {
      const path = joinPath(prefix, f.name);
      if (f.kind.case === "string") {
        const pattern = f.kind.value.pattern ?? "";
        if (pattern !== "" && !out.has(pattern)) {
          out.set(pattern, `${path}#pattern`);
        }
      }
      if (f.kind.case === "list") {
        walkFields(f.kind.value.items, `${path}[]`);
      }
      for (const child of nestedSchemas(f)) {
        walkFields(child.fields, path);
      }
    }
  };
  walkFields(s.fields, "");
  for (const [name, def] of Object.entries(s.defs)) {
    walkFields(def.fields, `$defs.${name}`);
  }
  return out;
}

// =============================================================================
// CEL <-> native conversion
// =============================================================================

function nativeToCel(x: Native): CelInput {
  if (x === null || typeof x === "boolean" || typeof x === "bigint" || typeof x === "number") {
    return x;
  }
  if (typeof x === "string" || x instanceof Uint8Array) {
    return x;
  }
  if (isMessage(x, DurationSchema) || isMessage(x, TimestampSchema)) {
    return x;
  }
  if (Array.isArray(x)) {
    return x.map(nativeToCel);
  }
  const out: Record<string, CelInput> = {};
  for (const [k, v] of Object.entries(x)) {
    out[k] = nativeToCel(v);
  }
  return out;
}

function celToNative(v: CelValue): Native {
  if (v === null || typeof v === "boolean" || typeof v === "bigint" || typeof v === "number") {
    return v;
  }
  if (typeof v === "string" || v instanceof Uint8Array) {
    return v;
  }
  if (isCelUint(v)) {
    return v.value;
  }
  if (isCelList(v)) {
    return [...v].map(celToNative);
  }
  if (isCelMap(v)) {
    const out: NativeStruct = {};
    for (const [k, el] of v) {
      const key = isCelUint(k) ? k.value.toString(10) : String(k);
      out[key] = celToNative(el);
    }
    return out;
  }
  if (isReflectMessage(v)) {
    const msg = v.message;
    if (isMessage(msg, DurationSchema) || isMessage(msg, TimestampSchema)) {
      return msg;
    }
    return null;
  }
  return null;
}

// =============================================================================
// AST dependency walk
// =============================================================================

function walkExpr(x: Expr, deps: string[]): void {
  const sel = selectPath(x);
  if (sel !== undefined) {
    if (sel !== "") {
      deps.push(sel);
    }
    return;
  }
  const kind = x.exprKind;
  switch (kind.case) {
    case "selectExpr":
      if (kind.value.operand !== undefined) {
        walkExpr(kind.value.operand, deps);
      }
      break;
    case "callExpr":
      if (kind.value.target !== undefined) {
        walkExpr(kind.value.target, deps);
      }
      for (const a of kind.value.args) {
        walkExpr(a, deps);
      }
      break;
    case "listExpr":
      for (const el of kind.value.elements) {
        walkExpr(el, deps);
      }
      break;
    case "structExpr":
      for (const entry of kind.value.entries) {
        if (entry.keyKind.case === "mapKey") {
          walkExpr(entry.keyKind.value, deps);
        }
        if (entry.value !== undefined) {
          walkExpr(entry.value, deps);
        }
      }
      break;
    case "comprehensionExpr": {
      const c = kind.value;
      for (const part of [c.iterRange, c.accuInit, c.loopCondition, c.loopStep, c.result]) {
        if (part !== undefined) {
          walkExpr(part, deps);
        }
      }
      break;
    }
    default:
      break;
  }
}

/** Resolves a select/index chain rooted at `root` to a dotted path. */
function selectPath(x: Expr): string | undefined {
  const kind = x.exprKind;
  switch (kind.case) {
    case "identExpr":
      return kind.value.name === "root" ? "" : undefined;
    case "selectExpr": {
      const operand = kind.value.operand;
      if (operand === undefined) {
        return undefined;
      }
      const base = selectPath(operand);
      if (base === undefined) {
        return undefined;
      }
      return base === "" ? kind.value.field : `${base}.${kind.value.field}`;
    }
    case "callExpr": {
      if (kind.value.function !== "_[_]" || kind.value.args.length !== 2) {
        return undefined;
      }
      const [operand, keyExpr] = kind.value.args;
      if (operand === undefined || keyExpr === undefined) {
        return undefined;
      }
      const base = selectPath(operand);
      if (base === undefined) {
        return undefined;
      }
      const constKind = keyExpr.exprKind;
      if (constKind.case !== "constExpr" || constKind.value.constantKind.case !== "stringValue") {
        return undefined;
      }
      const key = constKind.value.constantKind.value;
      return base === "" ? key : `${base}.${key}`;
    }
    default:
      return undefined;
  }
}

File: rust/src/engine.rs
//! The compiled engine: every CEL expression, regex pattern and Mustache
//! template compiled once, up front; conversions between the native model
//! and cel-interpreter values.

use std::collections::HashMap;
use std::sync::Arc;

use cel_interpreter::extractors::This;
use cel_interpreter::objects::{Key, Map as CelMap};
use cel_interpreter::{Context, Program, Value as Cel};

use crate::descriptor::{check_descriptor, join_path, nested_schemas, schema_err, SchemaError};
use crate::duration::{duration_from_nanos, duration_nanos};
use crate::formats::{core_formats, FormatRegistry};
use crate::gen::schemapb::schema::field::Kind as K;
use crate::gen::schemapb::{Schema, ValidationError};
use crate::render::{ascii_lower, ascii_upper};
use crate::value::{Native, NativeStruct, SchemaField};

/// An immutable compiled schema.
pub struct Engine {
    pub schema: Schema,
    pub formats: FormatRegistry,
    pub regexps: HashMap<String, regex::Regex>,
    programs: HashMap<String, Program>,
    asts: HashMap<String, cel_parser::Expression>,
    templates: HashMap<String, mustache::Template>,
}

/// Compiles a schema; `SchemaError` on any defect.
pub fn compile(schema: Schema, formats: FormatRegistry) -> Result<Engine, SchemaError> {
    let mut errs: Vec<ValidationError> = check_descriptor(&schema);

    let mut registry = core_formats();
    registry.extend(formats);

    let mut programs = HashMap::new();
    let mut asts = HashMap::new();
    for (src, path) in schema_exprs(&schema) {
        match Program::compile(&src) {
            Ok(p) => {
                if let Ok(ast) = cel_parser::parse(&src) {
                    asts.insert(src.clone(), ast);
                }
                programs.insert(src, p);
            }
            Err(e) => errs.push(schema_err(&path, &format!("cel: {e}"))),
        }
    }

    let mut regexps = HashMap::new();
    for (pattern, path) in schema_patterns(&schema) {
        match regex::Regex::new(&pattern) {
            Ok(re) => {
                regexps.insert(pattern, re);
            }
            Err(e) => errs.push(schema_err(&path, &format!("pattern: {e}"))),
        }
    }

    let mut templates = HashMap::new();
    for (name, src) in &schema.templates {
        match mustache::compile_str(src) {
            Ok(t) => {
                templates.insert(name.clone(), t);
            }
            Err(e) => errs.push(schema_err(
                &format!("templates.{name}"),
                &format!("mustache: {e}"),
            )),
        }
    }

    let engine = Engine {
        schema,
        formats: registry,
        regexps,
        programs,
        asts,
        templates,
    };
    if errs.is_empty() {
        errs.extend(engine.check_computed_cycles());
    }
    if errs.is_empty() {
        Ok(engine)
    } else {
        Err(SchemaError {
            result: crate::gen::schemapb::ValidationResult { errors: errs },
        })
    }
}

impl Engine {
    /// Runs a precompiled CEL plan (a sandboxed expression language — not
    /// Rust code evaluation). `Err(message)` on evaluation failure.
    pub fn eval(
        &self,
        src: &str,
        this: &Native,
        root: &NativeStruct,
        index: Option<i64>,
    ) -> Result<Native, String> {
        let Some(program) = self.programs.get(src) else {
            return Err(format!("expression not compiled: {src}"));
        };
        let mut ctx = Context::default();
        ctx.add_variable_from_value("this", native_to_cel(this));
        ctx.add_variable_from_value("root", native_to_cel(&Native::Struct(root.clone())));
        ctx.add_variable_from_value("index", Cel::Int(index.unwrap_or(0)));
        register_string_ext(&mut ctx);
        match program.execute(&ctx) {
            Ok(v) => cel_to_native(&v).ok_or_else(|| "unsupported CEL result".to_owned()),
            Err(e) => Err(e.to_string()),
        }
    }

    pub fn eval_bool(&self, src: &str, root: &NativeStruct) -> Result<bool, String> {
        match self.eval(src, &Native::Null, root, None)? {
            Native::Bool(b) => Ok(b),
            other => Err(format!("expression yields {other:?}, want bool")),
        }
    }

    /// The dotted root paths an expression reads.
    #[must_use]
    pub fn expr_deps(&self, src: &str) -> Vec<String> {
        let mut deps = Vec::new();
        if let Some(ast) = self.asts.get(src) {
            walk_deps(ast, &mut deps);
        }
        deps
    }

    #[must_use]
    pub fn render_template(
        &self,
        name: &str,
        ctx: &crate::render::RenderContext,
    ) -> Option<String> {
        let tmpl = self.templates.get(name)?;
        let mut out = Vec::new();
        tmpl.render(&mut out, ctx).ok()?;
        String::from_utf8(out).ok()
    }

    fn check_computed_cycles(&self) -> Vec<ValidationError> {
        let computed: HashMap<&str, &SchemaField> = self
            .schema
            .fields
            .iter()
            .filter(|f| matches!(f.kind.as_ref(), Some(K::Computed(_))))
            .map(|f| (f.name.as_str(), f))
            .collect();
        if computed.is_empty() {
            return Vec::new();
        }
        let deps: HashMap<&str, Vec<String>> = computed
            .iter()
            .map(|(name, f)| {
                let expr = match f.kind.as_ref() {
                    Some(K::Computed(c)) => c.expr.as_str(),
                    _ => "",
                };
                let ds = self
                    .expr_deps(expr)
                    .into_iter()
                    .filter(|d| d != name && computed.contains_key(d.as_str()))
                    .collect();
                (*name, ds)
            })
            .collect();

        fn visit<'a>(
            n: &'a str,
            deps: &'a HashMap<&str, Vec<String>>,
            color: &mut HashMap<&'a str, u8>,
        ) -> bool {
            match color.get(n) {
                Some(1) => return false,
                Some(2) => return true,
                _ => {}
            }
            color.insert(n, 1);
            for d in deps.get(n).map(Vec::as_slice).unwrap_or_default() {
                let Some((key, _)) = deps.get_key_value(d.as_str()) else {
                    continue;
                };
                if !visit(key, deps, color) {
                    return false;
                }
            }
            color.insert(n, 2);
            true
        }

        let mut color = HashMap::new();
        let mut errs = Vec::new();
        let mut names: Vec<&str> = computed.keys().copied().collect();
        names.sort_unstable();
        for name in names {
            if color.get(name) != Some(&2) && !visit(name, &deps, &mut color) {
                errs.push(schema_err(name, "computed field cycle"));
            }
        }
        errs
    }
}

/// The strings-extension subset (cel-interpreter has no built-in strings
/// extension; the remainder is a tracked gap).
fn register_string_ext(ctx: &mut Context) {
    ctx.add_function("lowerAscii", |This(s): This<Arc<String>>| ascii_lower(&s));
    ctx.add_function("upperAscii", |This(s): This<Arc<String>>| ascii_upper(&s));
    ctx.add_function("trim", |This(s): This<Arc<String>>| s.trim().to_owned());
}

// =============================================================================
// Expression / pattern collection
// =============================================================================

fn schema_exprs(s: &Schema) -> Vec<(String, String)> {
    let mut out: Vec<(String, String)> = Vec::new();
    let mut seen = std::collections::HashSet::new();
    let mut add = |src: &Option<String>, path: String, out: &mut Vec<(String, String)>| {
        if let Some(src) = src.as_deref() {
            if !src.is_empty() && seen.insert(src.to_owned()) {
                out.push((src.to_owned(), path));
            }
        }
    };

    #[allow(clippy::type_complexity)] // one-off callback used only here
    fn walk(
        fields: &[SchemaField],
        prefix: &str,
        add: &mut dyn FnMut(&Option<String>, String, &mut Vec<(String, String)>),
        out: &mut Vec<(String, String)>,
    ) {
        for f in fields {
            let path = join_path(prefix, &f.name);
            add(&f.when, format!("{path}#when"), out);
            add(&f.normalize, format!("{path}#normalize"), out);
            for (i, r) in f.rules.iter().enumerate() {
                add(&Some(r.expr.clone()), format!("{path}#rule[{i}]"), out);
            }
            match f.kind.as_ref() {
                Some(K::Computed(c)) => add(&Some(c.expr.clone()), format!("{path}#computed"), out),
                Some(K::Choice(ch)) => add(&ch.options_expr, format!("{path}#options"), out),
                Some(K::List(l)) => {
                    add(&l.count_expr, format!("{path}#count"), out);
                    walk(&l.items, &format!("{path}[]"), add, out);
                }
                _ => {}
            }
            for child in nested_schemas(f) {
                for (i, r) in child.rules.iter().enumerate() {
                    add(&Some(r.expr.clone()), format!("{path}#rule[{i}]"), out);
                }
                walk(&child.fields, &path, add, out);
            }
        }
    }

    for (i, r) in s.rules.iter().enumerate() {
        add(&Some(r.expr.clone()), format!("#rule[{i}]"), &mut out);
    }
    walk(&s.fields, "", &mut add, &mut out);
    for (name, def) in &s.defs {
        for (i, r) in def.rules.iter().enumerate() {
            add(
                &Some(r.expr.clone()),
                format!("$defs.{name}#rule[{i}]"),
                &mut out,
            );
        }
        walk(&def.fields, &format!("$defs.{name}"), &mut add, &mut out);
    }
    out
}

fn schema_patterns(s: &Schema) -> Vec<(String, String)> {
    let mut out = Vec::new();
    let mut seen = std::collections::HashSet::new();

    fn walk(
        fields: &[SchemaField],
        prefix: &str,
        seen: &mut std::collections::HashSet<String>,
        out: &mut Vec<(String, String)>,
    ) {
        for f in fields {
            let path = join_path(prefix, &f.name);
            if let Some(K::String(st)) = f.kind.as_ref() {
                if let Some(p) = st.pattern.as_deref() {
                    if !p.is_empty() && seen.insert(p.to_owned()) {
                        out.push((p.to_owned(), format!("{path}#pattern")));
                    }
                }
            }
            if let Some(K::List(l)) = f.kind.as_ref() {
                walk(&l.items, &format!("{path}[]"), seen, out);
            }
            for child in nested_schemas(f) {
                walk(&child.fields, &path, seen, out);
            }
        }
    }

    walk(&s.fields, "", &mut seen, &mut out);
    for (name, def) in &s.defs {
        walk(&def.fields, &format!("$defs.{name}"), &mut seen, &mut out);
    }
    out
}

// =============================================================================
// CEL <-> native conversion
// =============================================================================

fn native_to_cel(x: &Native) -> Cel {
    match x {
        Native::Null => Cel::Null,
        Native::Bool(b) => Cel::Bool(*b),
        Native::Int(n) => Cel::Int(*n),
        Native::Uint(n) => Cel::UInt(*n),
        Native::Double(f) => Cel::Float(*f),
        Native::Str(s) => Cel::String(Arc::new(s.clone())),
        Native::Bytes(b) => Cel::Bytes(Arc::new(b.clone())),
        Native::Duration(d) => Cel::Duration(chrono::Duration::nanoseconds(
            i64::try_from(duration_nanos(d)).unwrap_or(0),
        )),
        Native::Timestamp(t) => Cel::Timestamp(
            chrono::DateTime::from_timestamp(t.seconds, t.nanos.unsigned_abs())
                .map(Into::into)
                .unwrap_or_default(),
        ),
        Native::List(items) => Cel::List(Arc::new(items.iter().map(native_to_cel).collect())),
        Native::Struct(m) => {
            let map: HashMap<Key, Cel> = m
                .iter()
                .map(|(k, v)| (Key::String(Arc::new(k.clone())), native_to_cel(v)))
                .collect();
            Cel::Map(CelMap { map: Arc::new(map) })
        }
    }
}

fn cel_to_native(v: &Cel) -> Option<Native> {
    Some(match v {
        Cel::Null => Native::Null,
        Cel::Bool(b) => Native::Bool(*b),
        Cel::Int(n) => Native::Int(*n),
        Cel::UInt(n) => Native::Uint(*n),
        Cel::Float(f) => Native::Double(*f),
        Cel::String(s) => Native::Str(s.as_ref().clone()),
        Cel::Bytes(b) => Native::Bytes(b.as_ref().clone()),
        Cel::Duration(d) => Native::Duration(duration_from_nanos(i128::from(d.num_nanoseconds()?))),
        Cel::Timestamp(t) => Native::Timestamp(pbjson_types::Timestamp {
            seconds: t.timestamp(),
            nanos: i32::try_from(t.timestamp_subsec_nanos()).ok()?,
        }),
        Cel::List(items) => Native::List(
            items
                .iter()
                .map(cel_to_native)
                .collect::<Option<Vec<_>>>()?,
        ),
        Cel::Map(m) => {
            let mut out = NativeStruct::new();
            for (k, val) in m.map.iter() {
                let key = match k {
                    Key::String(s) => s.as_ref().clone(),
                    Key::Int(n) => n.to_string(),
                    Key::Uint(n) => n.to_string(),
                    Key::Bool(b) => b.to_string(),
                };
                out.insert(key, cel_to_native(val)?);
            }
            Native::Struct(out)
        }
        _ => return None,
    })
}

// =============================================================================
// AST dependency walk
// =============================================================================

fn walk_deps(expr: &cel_parser::Expression, deps: &mut Vec<String>) {
    use cel_parser::Expression as E;
    if let Some(path) = select_path(expr) {
        if !path.is_empty() {
            deps.push(path);
        }
        return;
    }
    match expr {
        E::Arithmetic(a, _, b) | E::Relation(a, _, b) => {
            walk_deps(a, deps);
            walk_deps(b, deps);
        }
        E::Ternary(c, a, b) => {
            walk_deps(c, deps);
            walk_deps(a, deps);
            walk_deps(b, deps);
        }
        E::Or(a, b) | E::And(a, b) => {
            walk_deps(a, deps);
            walk_deps(b, deps);
        }
        E::Unary(_, a) => walk_deps(a, deps),
        E::Member(target, member) => {
            walk_deps(target, deps);
            if let cel_parser::Member::Index(idx) = member.as_ref() {
                walk_deps(idx, deps);
            }
        }
        E::FunctionCall(f, target, args) => {
            walk_deps(f, deps);
            if let Some(t) = target {
                walk_deps(t, deps);
            }
            for a in args {
                walk_deps(a, deps);
            }
        }
        E::List(items) => {
            for a in items {
                walk_deps(a, deps);
            }
        }
        E::Map(entries) => {
            for (k, v) in entries {
                walk_deps(k, deps);
                walk_deps(v, deps);
            }
        }
        E::Atom(_) | E::Ident(_) => {}
    }
}

fn select_path(expr: &cel_parser::Expression) -> Option<String> {
    use cel_parser::{Atom, Expression as E, Member};
    match expr {
        E::Ident(name) => (name.as_ref() == "root").then(String::new),
        E::Member(target, member) => {
            let base = select_path(target)?;
            let key = match member.as_ref() {
                Member::Attribute(name) => name.as_ref().clone(),
                Member::Index(idx) => match idx.as_ref() {
                    E::Atom(Atom::String(s)) => s.as_ref().clone(),
                    _ => return None,
                },
                _ => return None,
            };
            Some(if base.is_empty() {
                key
            } else {
                format!("{base}.{key}")
            })
        }
        _ => None,
    }
}
